sdk/eventhub/azure-eventhubs/azure/eventhub/_eventprocessor/event_processor.py

```diff
@@ -0,0 +1,314 @@
+                    self._working_threads[partition_id] = threading.Thread(target=self._receive,
+                                                                           args=(partition_id, checkpoint))
+                    self._working_threads[partition_id].daemon = True
+                    self._threads_stop_flags[partition_id] = False
+                    self._working_threads[partition_id].start()
+                    log.info("Working thread started, ownership %r, checkpoint %r", partition_id, checkpoint)
+
+    @contextmanager
+    def _context(self, events):
+        # Tracing
+        span_impl_type = settings.tracing_implementation()  # type: Type[AbstractSpan]
+        if span_impl_type is None:
+            yield
+        else:
+            child = span_impl_type(name="Azure.EventHubs.process")
+            self._eventhub_client._add_span_request_attributes(child)  # pylint: disable=protected-access
+            child.kind = SpanKind.SERVER
+
+            for event in events:
+                event._trace_link_message(child)  # pylint: disable=protected-access
+            with child:
+                yield
+
+    def _receive(self, partition_id, checkpoint=None):  # pylint: disable=too-many-statements
+        log.info("start ownership %r, checkpoint %r", partition_id, checkpoint)
+        namespace = self._namespace
+        eventhub_name = self._eventhub_name
+        consumer_group_name = self._consumer_group_name
+        owner_id = self._id
+        checkpoint_offset = checkpoint.get("offset") if checkpoint else None
+        if checkpoint_offset:
+            initial_event_position = EventPosition(checkpoint_offset)
+        elif isinstance(self._initial_event_position, EventPosition):
+            initial_event_position = self._initial_event_position
+        elif isinstance(self._initial_event_position, dict):
+            initial_event_position = self._initial_event_position.get(partition_id, EventPosition("-1"))
+        else:
+            initial_event_position = EventPosition(self._initial_event_position)
+        if partition_id in self._partition_contexts:
+            partition_context = self._partition_contexts[partition_id]
+        else:
+            partition_context = PartitionContext(
+                namespace,
+                eventhub_name,
+                consumer_group_name,
+                partition_id,
+                owner_id,
+                self._partition_manager
+            )
+            self._partition_contexts[partition_id] = partition_context
+
+        partition_consumer = self._eventhub_client._create_consumer(  # pylint: disable=protected-access
+            consumer_group_name,
+            partition_id,
+            initial_event_position,
+            owner_level=self._owner_level,
+            track_last_enqueued_event_properties=self._track_last_enqueued_event_properties,
+            prefetch=self._prefetch,
+        )
+
+        def process_error(err):
```

Comment:
Why are we using nested functions here? It seems they're just being called directly below?
This pattern is used for both `close` and `process_error` but not for initialize?

I would prefer to see these three (close, error, initialize) moved into separate methods to simplify the code a little and improve testability.